LibJS: Fixes for array-like objects with large lengths

davidot opened the pull request:
This leaves array with:
```
test/built-ins/Array  2745/2763  ( 99.35%) [ ✅ 2745  ❌ 17💥️ 1]
```
but those are mostly this value problems.

This also removed the fast array paths in `push` and `pop` since these run into issues if `Array.prototype` is modified
We need to find some way of making these fast paths possible since this is much slower

davidot commented:
Oh that is a lot of tiny commits. Should I squash the last couple?

davidot commented on Userland/Libraries/LibJS/Runtime/PropertyName.h:
```diff
@@ -43,14 +43,20 @@ class PropertyName {
         VERIFY(index >= 0);
-        if constexpr (NumericLimits<T>::max() >= NumericLimits<u32>::max())
-            VERIFY(index < NumericLimits<u32>::max());
+        if constexpr (NumericLimits<T>::max() >= NumericLimits<u32>::max()) {
+            if (index >= NumericLimits<u32>::max()) {
+                m_string = String::number(index);
+                m_type = Type::String;
+                return;
+            }
+        }
+
+        m_type = Type::Number;
+        m_number = index;
```
Unsure if this is what we want but it is much easier on the eyes than: `PropertyName::from_value(global_object, Value(index))` everywhere

linusg replied:
yeah, that's good :+1:

linusg commented:
> Oh that is a lot of tiny commits. Should I squash the last couple?

IdanHo commented on Userland/Libraries/LibJS/Runtime/PropertyName.h:
```diff
@@ -43,14 +43,20 @@ class PropertyName {
-            VERIFY(index < NumericLimits<u32>::max());
+        if constexpr (NumericLimits<T>::max() >= NumericLimits<u32>::max()) {
+            if (index >= NumericLimits<u32>::max()) {
+                m_string = String::number(index);
+                m_type = Type::String;
+                return;
```
please also set m_may_be_number to false here, as we know it cant be.

IdanHo commented on Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp:
```diff
@@ -1367,18 +1367,24 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::includes)
     if (length == 0)
         return Value(false);
-    i32 from_index = 0;
+    ssize_t from_index = 0;
```
from_index is always non-negative here, why is it a ssize_t?

IdanHo commented on Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp:
```diff
@@ -1635,11 +1638,11 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::splice)
     } else if (vm.argument_count() >= 2) {
         insert_count = vm.argument_count() - 2;
-        i32 delete_count = vm.argument(1).to_i32(global_object);
+        double delete_count = vm.argument(1).to_integer_or_infinity(global_object);
```
no need to explicitly specify `double`, auto will do

IdanHo commented on Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp:
```diff
@@ -1676,14 +1679,15 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::splice)
-            if (!from.is_empty()) {
-                this_object->set(to, from, true);
+            if (from_present) {
+                auto from_value = this_object->get(i + actual_delete_count);
```
missing exception check

IdanHo commented on Userland/Libraries/LibJS/Runtime/PropertyName.h:
```diff
@@ -43,14 +43,21 @@ class PropertyName {
-        : m_type(Type::Number)
-        , m_number(index)
     {
         // FIXME: Replace this with requires(IsUnsigned<T>)?
         //        Needs changes in various places using `int` (but not actually being in the negative range)
         VERIFY(index >= 0);
```
This might be out of scope for your PR, but actually doing what this comments suggests will probably weed out more issues with using incorrect types for array indicies.

davidot replied:
I'll have a look what it finds

davidot replied:
Hmm not sure we want this as `PropertyName {1}` is no longer allowed and `PropertyName{0}` gets converted to a nullptr for char* I think

davidot replied:
It did find a number of places where we wrongly use `i32` so it might be something to look into

IdanHo replied:
yeah thats fine, you can leave it for a future PR

IdanHo reviewed (commented):
nice work, im not a fan of adding more size_t and ssize_t usages, since neither are correct on 64 bit systems (We should be using explicit u64 for lengths and u32 for indicies), but thats out of scope for this PR.

About the fast path issue: its not that bad since were missing a fast path for the majority of array methods right now anyways, but if we do decide to add them back, we can simply override `internal_set_prototype_of` in `Array`, set a boolean to true if its set to the non-default prototype, and then just add a helper method to array that checks that boolean is false and the length writable boolean is true, and afaik that should cover our bases.

davidot commented:
I did not make the `Integral` -> `Unsigned` change in PropertyName see:

Not sure we want this as `PropertyName {1}` is no longer allowed and `PropertyName{0}` gets converted to a `nullptr` for the  `char*` constructor I think
It did find a number of places where we wrongly use i32 so it might be something to look into

I fixed the issues and moved from size_t and ssize_t to u64 and i64 respectively in the methods I changed in this PR

IdanHo commented:
We might want to add a `Value(u64)` constructor that VERIFYs that the value is `< 2^53 - 1`

linusg reviewed (commented):
Some notes about commit messages:
- Please write full sentences, including punctuation - that includes a trailing `.` :)
- Please follow the `If applied, this commit will... <insert message>` scheme for the commit title (last commit)

IdanHo commented:
`as_object` instead of `to_object` in the last commit message
